keys/CipherKeyGenerator.java

```diff
@@ -0,0 +1,26 @@
+package keys;
+
+import java.util.HashMap;
+
+public abstract class CipherKeyGenerator {
+  protected HashMap<Character, Character> cipherKey;
+
+  public CipherKeyGenerator() {
+    cipherKey = new HashMap<>();
+  }
+
+  public abstract void generateKey(int shift, char[] charSet);
+
+  protected int reduceShiftToCharsetRange(int shift, int charSetLength) {
+    int adjustedShift = shift;
+
+    while (adjustedShift >= charSetLength) {
+      adjustedShift -= charSetLength;
+    }
+    return adjustedShift;
+  }
+
+  public HashMap<Character, Character> getKey() {
```

Comment:
Check the Polyphormism. Im sure you need public Map<... here not HashMap
https://www.javatpoint.com/runtime-polymorphism-in-java